File: TypeScript/BinarySearchTree.ts
class TreeNode<T> {
  value: T;
  left: TreeNode<T> | null;
  right: TreeNode<T> | null;

  constructor(value: T) {
    this.value = value;
    this.left = null;
    this.right = null;
  }

  depth(): number {
    const leftDepth = this.left?.depth() || 0;
    const rightDepth = this.right?.depth() || 0;

    return 1 + Math.max(leftDepth, rightDepth);
  }

  push(value: T): TreeNode<T> {
    if (value < this.value) {
      if (this.left === null) this.left = new TreeNode(value);
      else this.left.push(value);
    } else if (value > this.value) {
      if (this.right === null) this.right = new TreeNode(value);
      else this.right.push(value);
    }

    return this;
  }

  find(fn: (value: T) => boolean): T | null {
    if (fn(this.value)) return this.value;

    let leftFind = this.left ? this.left.find(fn) : null;
    if (leftFind) return leftFind;

    let rightFind = this.right ? this.right.find(fn) : null;
    if (rightFind) return rightFind;

    return null;
  }

  contains(value: T): boolean {
    if (value < this.value) {
      if (!this.left) return false;
      return this.left.contains(value);
    } else if (value > this.value) {
      if (!this.right) return false;
      return this.right.contains(value);
    }

    return true;
  }

  map(fn: (value: T) => any): TreeNode<any> {
    let newNode = new TreeNode(fn(this.value));

    if (this.left) newNode.left = this.left.map(fn);
    if (this.right) newNode.right = this.right.map(fn);

    return newNode;
  }
}

export class BinarySearchTree<T> {
  root: TreeNode<T> | null;

  constructor() {
    this.root = null;
  }

  isEmpty(): boolean {
    return this.root === null;
  }

  depth(): number {
    if (!this.root) return 0;

    return this.root.depth();
  }

  push(value: T): BinarySearchTree<T> {
    if (!this.root) this.root = new TreeNode(value);
    else this.root.push(value);

    return this;
  }

  find(fn: (value: T) => boolean): T | null {
    if (!this.root) return null;

    return this.root.find(fn);
  }

  contains(value: T): boolean {
    if (!this.root) return false;

    return this.root.contains(value);
  }

  map(fn: (value: T) => any): BinarySearchTree<any> {
    if (!this.root) return new BinarySearchTree();

    let newTree = new BinarySearchTree();
    newTree.root = this.root.map(fn);

    return newTree;
  }
}

File: Cpp/BinarySearchTree.cpp
#include "BinarySearchTree.h"

template <typename T>
TreeNode<T>::TreeNode(T value) {
  this->_value = value;
  this->_left = nullptr;
  this->_right = nullptr;
}

template <typename T>
int TreeNode<T>::depth() {
  int leftDepth = !_left ? 0 : _left->depth();
  int rightDepth = !_right ? 0 : _right->depth();

  return 1 + max(leftDepth, rightDepth);
}

template <typename T>
void TreeNode<T>::push(T value) {
  if (value < _value) {
    if (!_left)
      _left = new TreeNode(value);
    else
      _left->push(value);
  } else if (value > _value) {
    if (!_right)
      _right = new TreeNode(value);
    else
      _right->push(value);
  }
}

template <typename T>
bool TreeNode<T>::contains(T value) {
  if (value < _value) {
    if (!_left) return false;
    return _left->contains(value);
  } else if (value > _value) {
    if (!_right) return false;
    return _right->contains(value);
  }

  return true;
}

template <typename T>
template <typename U>
TreeNode<U>* TreeNode<T>::map(U (*fn)(T)) {
  auto newNode = new TreeNode(fn(_value));

  if (_left) newNode->_left = _left->map(fn);
  if (_right) newNode->_right = _right->map(fn);

  return newNode;
}

template <typename T>
T* TreeNode<T>::find(bool (*fn)(T)) {
  if (fn(_value)) return &_value;

  auto leftFind = _left ? _left->find(fn) : nullptr;
  if (leftFind) return leftFind;

  auto rightFind = _right ? _right->find(fn) : nullptr;
  if (rightFind) return rightFind;

  return nullptr;
}

template <typename T>
BinarySearchTree<T>::BinarySearchTree() {
  _root = nullptr;
}

template <typename T>
bool BinarySearchTree<T>::isEmpty() {
  return !_root;
}

template <typename T>
int BinarySearchTree<T>::depth() {
  if (!_root) return 0;

  return _root->depth();
}

template <typename T>
void BinarySearchTree<T>::push(T value) {
  if (!_root)
    _root = new TreeNode<T>(value);
  else
    _root->push(value);
}

template <typename T>
bool BinarySearchTree<T>::contains(T value) {
  if (!_root) return false;

  return _root->contains(value);
}

template <typename T>
template <typename U>
BinarySearchTree<U>* BinarySearchTree<T>::map(U (*fn)(T)) {
  if (!_root) return new BinarySearchTree<U>();

  auto newTree = new BinarySearchTree<U>();
  newTree->_root = _root->map(fn);

  return newTree;
}

template <typename T>
T* BinarySearchTree<T>::find(bool (*fn)(T)) {
  if (!_root) return nullptr;

  return _root->find(fn);
}
